controller/dataplane/controller_reconciler_utils.go

```diff
@@ -77,6 +84,170 @@ func (r *Reconciler) ensureDataPlaneAddressesStatus(
+		return isForKPI
+	}), nil
+}
+
+func populateDedicatedConfigMapForKongPluginInstallation(
+	ctx context.Context,
+	logger logr.Logger,
+	c client.Client,
+	cms []corev1.ConfigMap,
+	kpiNN types.NamespacedName,
+	dataplane *operatorv1beta1.DataPlane,
+) (customPlugin, ctrl.Result, error) {
+	kpi, ready, err := verifyKPIReadinessForDataPlane(ctx, logger, c, dataplane, kpiNN)
+	if err != nil {
+		return customPlugin{}, ctrl.Result{}, err
+	}
+	if !ready {
+		return customPlugin{}, ctrl.Result{Requeue: true}, nil
+	}
+
+	var underlyingCM corev1.ConfigMap
+	backingCMNN := types.NamespacedName{
+		Namespace: kpi.Namespace,
+		Name:      kpi.Status.UnderlyingConfigMapName,
+	}
+	log.Trace(logger, fmt.Sprintf("Fetch underlying ConfigMap %s for KongPluginInstallation", backingCMNN), kpi)
+	if err := c.Get(ctx, backingCMNN, &underlyingCM); err != nil {
+		return customPlugin{}, ctrl.Result{}, fmt.Errorf("could not fetch underlying ConfigMap to clone %s: %w", backingCMNN, err)
+	}
+
+	log.Trace(logger, "Find ConfigMap mapped to KongPluginInstallation", kpi)
+	mappedConfigMapForKPI := lo.Filter(cms, func(cm corev1.ConfigMap, _ int) bool {
+		kpiNN := cm.Annotations[consts.AnnotationMappedToKongPluginInstallation]
+		return kpiNN == client.ObjectKeyFromObject(&kpi).String()
+	})
+	var cm corev1.ConfigMap
+	switch len(mappedConfigMapForKPI) {
+	case 0:
+		log.Trace(logger, "Create new ConfigMap for KongPluginInstallation", kpi)
+		cm.GenerateName = dataplane.Name + "-"
+		cm.Namespace = dataplane.Namespace
+		k8sutils.SetOwnerForObject(&cm, dataplane)
+		k8sresources.LabelObjectAsDataPlaneManaged(&cm)
+		k8sresources.AnnotateConfigMapWithKongPluginInstallation(&cm, kpi)
+		cm.Data = underlyingCM.Data
+		if err := c.Create(ctx, &cm); err != nil {
+			return customPlugin{}, ctrl.Result{}, fmt.Errorf("could not create new ConfigMap for KongPluginInstallation: %w", err)
+		}
+	case 1:
+		cm = mappedConfigMapForKPI[0]
+		log.Trace(logger, fmt.Sprintf("Update existing ConfigMap %s for KongPluginInstallation", client.ObjectKeyFromObject(&cm)), kpi)
+		cm.Data = underlyingCM.Data
+		if err := c.Update(ctx, &cm); err != nil {
+			if k8serrors.IsConflict(err) {
+				return customPlugin{}, ctrl.Result{Requeue: true}, nil
+			}
+			return customPlugin{}, ctrl.Result{}, fmt.Errorf("could not update mapped: %w", err)
+		}
+	default:
+		// It should never happen.
+		names := strings.Join(lo.Map(mappedConfigMapForKPI, func(cm corev1.ConfigMap, _ int) string {
+			return client.ObjectKeyFromObject(&cm).String()
+		}), ", ")
+		return customPlugin{}, ctrl.Result{}, fmt.Errorf("unexpected error happened - more than one ConfigMap found: %s", names)
+	}
+	return customPlugin{
+		Name:        kpi.Name,
+		ConfigMapNN: client.ObjectKeyFromObject(&cm),
+		Generation:  kpi.Generation,
+	}, ctrl.Result{}, nil
+}
+
+// verifyKPIReadinessForDataPlane Reports in DataPlane status when KPI does not exist or it
```

Comment:
```suggestion
// verifyKPIReadinessForDataPlane reports in DataPlane status when KPI does not exist or it
```

Reply:
Nit: reports feels vague here. Reading this godoc I don't know what are going to be the side effects of this `func`. Looking at the source code it seems that `DataPlane`'s conditions are updated according to the state of KPI. Can we mention that in the comment?

Reply:
I reworded it entirely, PTAL